network: IPv4LL and ACD fixes

yuwata opened the pull request:
Fixes #12145.
Fixes #17235.

yuwata commented:
I added this to v249 milestone. If this is too late to be included in v249, feel free to drop it. But this is slightly large, and maybe hard to backport this after v249.

poettering commented on src/libsystemd-network/sd-ipv4acd.c:
```diff
@@ -432,12 +432,24 @@ const char *sd_ipv4acd_get_ifname(sd_ipv4acd *acd) {
+        if (!sd_ipv4acd_is_running(acd))
+                return 0;
+
+        assert(acd->fd);
```
`assert(acd->fd >= 0)`

poettering commented on src/libsystemd-network/sd-ipv4acd.c:
```diff
@@ -490,14 +490,35 @@ int sd_ipv4acd_set_callback(sd_ipv4acd *acd, sd_ipv4acd_callback_t cb, void *use
+        if (!sd_ipv4acd_is_running(acd))
+                return 0;
+
+        assert(acd->fd);
```
>= 0

poettering commented on src/libsystemd-network/sd-ipv4acd.c:
```diff
@@ -316,6 +305,41 @@ static int ipv4acd_on_timeout(sd_event_source *s, uint64_t usec, void *userdata)
+                return false; /* target IP address does not match, BPF code is broken? */
+
+        SET_FOREACH(a, acd->other_mac_addrs)
+                if (memcmp(arp->arp_sha, a, ETH_ALEN) == 0)
```
ether_addr_equal()?

poettering replied:
ah, i see, arp_sha is not of type struct ether_addr. ignore hence

poettering commented on src/libsystemd-network/sd-ipv4acd.c:
```diff
@@ -453,6 +455,28 @@ int sd_ipv4acd_set_mac(sd_ipv4acd *acd, const struct ether_addr *addr) {
+
+        free(set_remove(acd->other_mac_addrs, addr));
+        return 0;
+}
```
i wonder if this shouldn't be done with a callback instead. I.e. instead of propagating the MAC addresses of the whole system continously into all sd_ipv4acd objects, do it the other way round: allow configuring a callback function that is called whenver a MAC address should be checked if local. And then maintain a single global hashmap inside of networkd where we keep track of all local MAC addreses and the interfaces they belong to.

This could even be useful to detect MAC address collisions when automatically assigning a MAC address or so.

Or in other words:

1. Extend the `Manager` object to gain a new field `links_by_mac_address`. It's keyed by the mac address and points to the interface with that MAC address, as is kept track of the same way as `links_by_name` is. (if we want to support that the same MAC address may be set for multiple local interfaces, then I guess we need to make sure that links with the same one are linked up in a linked list, and the `links_by_mac_address` hashmap points only to the first of them.

2. Add sd_ipv4acd_set_other_mac_callback() that takes an `sd_ipv4acd_other_mac_callback_t` parameter which is a function that returns 0, 1 or error, and gets a `const struct ether_addr` as parameter.

3. Add glue code to networkd to implement sd_ipv4acd_other_mac_callback_t once, and make it check the `links_by_mac_address` hashmap.

I think this would be more robust, less duplication and use less resources.

(Btw, I think it would be nice to rename the `links` field in `Manager` to `links_by_ifindex`, so that we then nicely have `links_by_ifindex`, `links_by_ifname` and `links_by_mac`)

yuwata commented:
@poettering Thank you for your review. Now, the callback mechanisms which checks the sender hardware address is introduced to sd-ipv4acd and sd-ipv4ll. PTAL.

keszybz reviewed (commented):
LGTM. I'd be inclined to merge this for 249.

keszybz commented on src/network/networkd-address.c:
```diff
@@ -1153,6 +1153,22 @@ int link_request_address(
+        if (r < 0)
+                return r;
+
+        req->after_configure = static_address_after_configure;
```
@yuwata ^

yuwata replied:
Thank you for catching this. This is an error. Fixed.

yuwata commented on src/network/networkd-ipv4acd.c:
```diff
@@ -0,0 +1,234 @@
+                                               IPV4_ADDRESS_FMT_VAL(address->in_addr.in));
+                        link_enter_failed(link);
+                }
+                break;
```
You mean about the log messages? The messages in the callback function is updated. PTAL.

yuwata commented:
@keszybz Thank you for your review. Force-pushed a revised version. In addition to the above points you commented, the following changes are applied.
- in dhcp_lease_lost(), now the pending request is also dropped.
- in on_acd(), DHCPv4 address is not removed in the function, but dhcp_lease_lost() drop it.
- split on_acd() into several small functions.

keszybz commented on src/network/networkd-ipv4acd.c:
```diff
@@ -0,0 +1,271 @@
+                /* Unlikely, but during probing the address, the lease may be lost. */
+                return 0;
+
+        log_link_warning(link, "Dropping DHCPv4 lease, as an address conflict is detected.");
```
"is" → "was"

keszybz reviewed (commented):
s390x: "boot-and-services": `test_service (__main__.NspawnTest) ... Failed to dump process list for 'systemd-nspawn@c1.service', ignoring: Unit systemd-nspawn@c1.service not loaded.`
Looks unrelated.

LGTM, let's get this merged.
